scripts/mkcap.py

```diff
@@ -270,62 +296,128 @@ def print_def_intent(self, metadata):
         dictionary to resolve lower bounds for array dimensions.'''
         # Resolve dimensisons to local names using undefined upper bounds (assumed shape)
         dimstring = self.dimstring_local_names(metadata, assume_shape = True)
+        # It is an error for host model variables to have the optional attribute in the metadata
```

Comment:
I don't see where you could have put it. 
What exactly is the rule here? I thought the host model is the source of the need for optional variables. That is, if a host model can choose to not allocate one of its variables, it needs to be declared optional by routines that use it.

Is it possible that the distinction is really about not allowing the `optional` attribute for `host` or `module` metadata tables? This could easily be done in metavar.py by moving the `VariableProperty`, `optional` from the `__spec_props` list to the `__var_props` list. Since `HostModel` only allows 'module' and 'host' metadata tables, I think this would act as the correct barrier.

I'm not sure where this leaves DDTs. For a host DDT, can an element have the `active` attribute? What happens if you try that with a DDT that is used in a scheme? We may need to add logic here as well.

Thoughts?

Reply:
Yes, this is about the metadata only. The rule is quite simple (if I understood it correctly):
1. host models can have the active attribute for a variable or not (default value is `true` = always allocated)
2. schemes can have the optional attribute for a variable or not (default value is `false` = always allocated by the host model)
3. it is an error of a host model has `optional` in the metadata attribute
4. it is an error of a scheme has `active` in the metadata attribute
5. if a host model has an active attribute with a value other than `true`, any scheme using that environment variable must use `optional = true` (requirement3)
6. the opposite is not true or at least should not be true: a scheme can have `optional = true`, because model A requires that, but there can be a model B that doesn't declare it as `active` in its metadata.

**prebuild** checks for 1-5 in its own layer after receiving the metadata from capgen. This is because we only harvest information on host metadata and scheme metadata separately (except for passing known host DDTs) and do not rely on the metadata comparison host <--> model from capgen. I will note that at this point, prebuild spits out a warning for 6 if it finds a variable with optional being true but no non-default active attribute in the model. I used this to clean up the metadata, knowing that the current ccpp physics are tied to the GFS data structures used by UFS, SCM, NEPTUNE.

Regarding your pointer question. That gets tricky. The only way out I can see here is that if a host ddt has an active attribute other than the default true, all it's members (if declared explicitly in the metadata) need that to have active to, with the allocation condition being that of the DDT (if the member is always allocated when the DDT is allocated) OR a Fortran logical && of the DDTs allocation condition and its own. We need to be explicit here and have both allocation conditions in the member var's active attribute instead of implicitly inheriting it from the DDT. That would greatly confuse the physics/model developers and users.

Reply:
1,2,5, and 6 are applied In the Capgen implementation, with the exception of 3 and 4. I will add these soon, and revisit the DDTs.